source/Irrlicht/COpenGLDriver.cpp

```diff
@@ -2239,7 +2278,12 @@ void COpenGLDriver::SAuxContext::flushStateGraphics(uint32_t stateBits)
 			const COpenGLSpecializedShader* shdr = static_cast<const COpenGLSpecializedShader*>(currentState.pipeline.graphics.pipeline->getShaderAtIndex(i));
 			//assert(shdr); //this would be weird
 
-			shdr->setUniformsImitatingPushConstants(pushConstantsState[EPBP_GRAPHICS].data, this->ID);
+            //WARNING order of calls getShaderGLnameForCtx() and getUniformLocationsForStage() is important!!
```

Comment:
why?

Reply:
because `getShaderGLnameForCtx` might create the GL name and `getUniformLocationsForStage` will need GL name to query locations (ofc locations are cached once queried, but order of those calls are still important for the first time when locations are yet unknown)

Reply:
> because getShaderGLnameForCtx might create the GL name

We could just create all the thread's programs (all the duplicates) at once during pipeline creation so we never, ever, stutter during a frame.

> getUniformLocationsForStage will need GL name to query locations (ofc locations are cached once queried, but order of those calls are still important for the first time when locations are yet unknown)

we should always compile/create at least one shader program upon pipeline creation and hence the uniform locations should be queried at that time and stored in the specialized shader (they will be identical for all following copies).

To avoid any hitches during rendering.

Reply:
done

Reply:
actually only the 2nd one
So u want to create for all threads at once? Now pipeline irrbaw objects store binaries and create GL names from it when a thread wants to use the pipeline

Reply:
there's probably an overhead in creating a new shader even from a binary, so maybe doing all up-front is a better idea.